homeassistant/components/templated_automation/__init__.py

```diff
@@ -0,0 +1,106 @@
+https://home-assistant.io/components/templated_automation/
+"""
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.binary_sensor import \
+    DOMAIN as BINARY_SENSOR_DOMAIN
+from homeassistant.components.input_boolean import \
+    DOMAIN as INPUT_BOOLEAN_DOMAIN
+from homeassistant.components.light import DOMAIN as LIGHT_DOMAIN
+from homeassistant.components.switch import DOMAIN as SWITCH_DOMAIN
+from homeassistant.const import CONF_SOURCE, STATE_OFF, \
+    SERVICE_TURN_OFF, SERVICE_TURN_ON, ATTR_ENTITY_ID
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change
+
+CONF_BIDIRECTIONAL = 'bidirectional'
+CONF_INVERTED = 'inverted'
+CONF_TARGET = 'target'
+
+DEFAULT_BIDIRECTIONAL = False
+DEFAULT_INVERTED = False
+
+DOMAIN = 'templated_automation'
+
+BINARY_ENTITY_DOMAINS = [
+    BINARY_SENSOR_DOMAIN,
+    SWITCH_DOMAIN,
+    LIGHT_DOMAIN,
+    INPUT_BOOLEAN_DOMAIN
+]
+
+
+def binary_entity_id(value: str) -> str:
+    """Validate that an Entity ID is from a binary domain."""
+    parts = value.split('.', 1)
+    if parts[0] not in BINARY_ENTITY_DOMAINS \
+            and parts[1] not in BINARY_ENTITY_DOMAINS:
+        raise vol.Invalid(
+            'Entity ID {} is not from a binary entity domain'.format(value)
+        )
+    return value
+
+
+AUTOMATION_TEMPLATE = vol.Schema({
+    vol.Required(CONF_SOURCE): vol.All(cv.entity_id, binary_entity_id),
+    vol.Required(CONF_TARGET): vol.All(cv.entity_id, binary_entity_id),
+    vol.Optional(
+        CONF_BIDIRECTIONAL, default=DEFAULT_BIDIRECTIONAL
+    ): cv.boolean,
+    vol.Optional(CONF_INVERTED, default=DEFAULT_INVERTED): cv.boolean
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [AUTOMATION_TEMPLATE])
+}, extra=vol.ALLOW_EXTRA)
+
+ATTR_NEW_STATE = 'new_state'
+
+
+async def async_setup(hass, config):
+    """Set up the Templated Automation component."""
+    automations = {}
+    for conf in config[DOMAIN]:
+        source = conf[CONF_SOURCE]
+        target = conf[CONF_TARGET]
+        inverse = conf.get(CONF_INVERTED)
+        automation = TemplatedAutomation(hass, source, target, inverse)
+        await automation._init()
+        automations[source] = automation
+
+    hass.data[DOMAIN] = automations
+    return True
+
+
+class TemplatedAutomation:
+    """Represents a Templated Automation."""
+
+    def __init__(self, hass, source, target, inverse):
+        """Initialize a Templated Automation object."""
+        self._hass = hass
+        self._source = source
+        self._target = target
+        self._inverse = inverse
+        self._unsub = None
+
+    async def _init(self):
+        """Add the callback for state changes."""
+        self._unsub = await async_track_state_change(
+            self._hass, self._source, self.handle_state_change
+        )
+
+    @callback
+    def handle_state_change(self, entity_id, from_s, to_s):
```

Comment:
`new_state` and `old_state` are more clear in my opinion.

Reply:
Make this a coroutine instead.

Reply:
Even using a coroutine, it blows up with this error:
```
2019-01-26 13:01:57 ERROR (MainThread) [homeassistant.setup] Error during setup of component templated_automation
Traceback (most recent call last):
  File "/mnt/d/rohan/Repos/home-assistant/homeassistant/setup.py", line 145, in _async_setup_component
    hass, processed_config)
  File "/mnt/d/rohan/Repos/home-assistant/homeassistant/components/templated_automation/__init__.py", line 73, in async_setup
    await automation._init()
  File "/mnt/d/rohan/Repos/home-assistant/homeassistant/components/templated_automation/__init__.py", line 94, in _init
    self._hass, self._source, self.handle_state_change
TypeError: object function can't be used in 'await' expression
```
and I think that's telling me that I can't have `handle_state_change` as part of the TemplatedAutomation object right?